cpmpy/expressions/globalconstraints.py

```diff
@@ -190,7 +190,7 @@ def __init__(self, *args):
         super().__init__("alldifferent_except0", flatlist(args))
 
     def decompose(self):
-        return [((var1 != 0) & (var2 != 0)).implies(var1 != var2) for var1, var2 in all_pairs(self.args)]
+        return [(var1 == var2).implies(var1 == 0) for var1, var2 in all_pairs(self.args)]
```

Comment:
is this true? Should it not be `((var1 == 0) | (var2 == 0))` ?

Reply:
basically the preferred decomposition is a single clause `(var1 == 0) | (var2 == 0) | (var1 != var2)` imho

Reply:
Well if `var1 == var2` the if `var1 == 0`, it also holds that `var2 == 0`. 
I double-checked and the new decomposition is indeed equivalent.

The new one requires much less auxiliary variables when flattening:
```python
flatten_constraint(old_decomp)
>>[((BV29) and (BV30)) -> (BV31),
    (a != 0) == (BV29),
    (b != 0) == (BV30),
    ((a) != (b)) == (BV31)
]

flatten_constraint(new_decomp)
>>[((a) == (b)) -> (BV32)
    (a == 0) == (BV32)]
```